### DISCUSSION TITLE
Uploading a changed boot.py yields no changes on ESP32

### DISCUSSION THREAD

**Discussion by BitHacker49** [micropython/micropython#15444] [ESP32]
Hi,

I'm brandnew to all that microcontroller stuff and currently trying to understand the boot process for offline operation of the ESP32.

I changed the **boot.py** file on the PC, deleted its old version on ESP and uploaded the changed boot.py to the ESP. Upload pretends to work (boot.py showing up in the "MicroPython device" window). But when I click on that boot.py on the ESP, Thonny shows me still the old boot code version - **before** the changes I made. Also, pressing the restart button on the ESP board yields the old boot behavour rather than the changed one. I never get the latest version, as prepared on the PC, working on the ESP. 

Also, downloading to PC the boot.py after changing it directely on ESP doesn't work. The downloading process pretends to work (even displays a correct warning about overwriting files), but the file on the PC remains the same.

A workaround could of course be to implement all changes twice - on the PC and in addition directly on the ESP. (And that seems to work.) But before accepting this as a solution I'd like to understanding first what happens and what am I doing wrong.

Thanks for any ideas.

**Comment by Jeff-Squared** [comment]
The Thonny "Files" view will show you the files in the working directory of your PC and below that a list of files stored on the flash of your device.

If you have a file in the working directory called boot.py and a file in the device's flash called boot.py. Although they have the same name and contents they are two separate files, so when you open and modify the file on the PC and hit the save icon the file in flash remains unchanged. You can alter this behaviour by clicking the "File" menu item and choosing "Save As" or "Save Copy"

Perhaps a better option is to open the file from the device flash and make modifications and saves only to that file while preserving the original on the PC, that way when you have finished you can copy to both locations or if you wish you can restore the original from the PC.

**Reply by BitHacker49** [reply]
Thanks for your quick response, Jeff.

> You can alter this behaviour by clicking the "File" menu item and choosing "Save As" or "Save Copy"

I thought I could achieve the same by right-clicking the file in the PC view and selecting "upload to \". But although that pretends to work, it actually doesn't (see above). Is that really how it's intended to work?

> Perhaps a better option is to open the file from the device flash and make modifications and saves only to that file while preserving the original on the PC.

Looks like that has to be my workaround.

To be fair - I should have mentioned that I'm still on Thonny 3.2.7, as this is the version recommended by the Freenove tutorial I'm working from. Since I also had tremendous problems with deinstalling Thonny, I'm hesitant to upgrade. So currently, Thonny is a real mess on my PC - not sure that I stay with it. 

By the way, for good reasons my working directory for this project is on OneDrive. Could this be a cause of my trouble?

**Comment by Jeff-Squared** [comment]
I had never thought of OneDrive as the working directory but I don't see why that would make a difference, I also use "upload" (PC to device) and "download (device to PC) without issue so I'm unsure what's going on there, my version is 4.1.4

I think working with the file on the device is the most prudent thing to do, I have lost files myself doing things the other way around.

Sorry I can't offer more help but I still think Thonny is a good IDE especially for beginners like me.

**Reply by BitHacker49** [reply]
Yeah, I think I should just gain some more experience before judging. In general I'm delighted as to how quick my very first "blink" project came into working.

Meanwhile I ran into some best practise advices as to how to deal with some known issues of Thonny running the ESP32. What I understood so far is that if you run the code from the ESP (with proper boot.py) you should preferably use the RST button on the board to start the program. That alone helped me a lot. Besides that, when navigating through files in PC and device windows (like I mentioned above), you have to **double**-click them to get them displayed. This might be one major reason for the issues I reported here.

Anyhow, thanks for your help, Jeff.
